Prevent to deep-copy nodes that have been stored. Fixes #241

bosonie opened the pull request:
Fixes #241

bosonie commented:
I believe that tests are not necessary since this feature is tested already when the various child classes of `InputGenerator` are tested. The relax workchain testes already inputs that are not nodes and unstored nodes. The bands workchain tests will test stored nodes.

sphuber reviewed (changes requested):
I think that we actually _do_ need tests here. What you describe being tested is that things _run_, but not that nodes are not accidentally _cloned_, which is actually happening in this implementation. You are only checking the top level namespace for nodes, but they could be nested deeply within. Those will still get deepcopied. So to do this properly, you need to make it recursive. But you need to do it in a backtracking manner. You will need to go all the way down the namespaces to find where the nodes are and clone deepcopy the dictionaries by hand and the mutable values within it, except for the nodes which you keep.

bosonie commented:
@sphuber I fixed the issue and added tests. The tests can be improved after we merge also the feature on bands. In that PR (#242) the `InputGenerator` is changed to not depend on `ProtocolRegistry`.
Ready for approval.

sphuber commented on tests/generators/test_generator.py:
```diff
@@ -0,0 +1,63 @@
+    Test the serializer of the arguments of `get_builder`
+    """
+
+    clas = test_inputgen(inputs_dict={'structure': orm.StructureData})
+
+    class_ins = clas(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+
+    in_struct = generate_structure(symbols=('Si',))
+    buil, args = class_ins.get_builder(structure=in_struct)  # pylint: disable=unused-variable
+    # in_struct is an unstored node, therefore it is deepcopied and
+    # for this reason the uuid of the initial structure and the one
+    # inside `get_builder` are different
+    assert args['structure'].uuid != in_struct.uuid
+
+    in_struct = generate_structure(symbols=('Si',)).store()
+    buil, args = class_ins.get_builder(structure=in_struct)  # pylint: disable=unused-variable
+    # in_struct is now stored, therefore it is maintained
+    # within `get_builder`
+    assert args['structure'].uuid == in_struct.uuid
```
Removed the fact that the `get_builder` also returns the arguments as that is changing the signature and is not actually necessary. We want to check that the `kwargs` that are passed to it are not modified, so we can check that directly.
```suggestion
    cls = generate_input_generator_cls(inputs_dict={'structure': orm.StructureData})
    generator = cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))

    structure = generate_structure(symbols=('Si',))
    kwargs = {'structure': structure}
    generator.get_builder(**kwargs)
    # structure is an unstored node, therefore it is deepcopied and so the UUID should change.
    assert kwargs['structure'].uuid != structure.uuid

    structure = generate_structure(symbols=('Si',)).store()
    kwargs = {'structure': structure}
    generator.get_builder(**kwargs)
    # The structure is now stored and so should be the same.
    assert kwargs['structure'].uuid == structure.uuid
```

bosonie replied:
I am not totally sure about this @sphuber. I wanted to check that the kwargs are deepcopied inside the `get_builder` if they are not stored nodes. Are you instead proposing to change one of the  kwargs (for instance structure) inside the `get_builder` and see that the original did not change??

sphuber commented on tests/generators/test_generator.py:
```diff
@@ -0,0 +1,63 @@
+    """
+    Test the serializer of the arguments of `get_builder` in case of input_namespaces
+    """
+
+    clas = test_inputgen(inputs_dict={'space.structure': orm.StructureData}, namespaces=['space'])
+
+    class_ins = clas(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+
+    in_struct = generate_structure(symbols=('Si',))
+    buil, args = class_ins.get_builder(space={'structure': in_struct})  # pylint: disable=unused-variable
+    # in_struct is an unstored node, therefore it is deepcopied and
+    # for this reason the uuid of the initial structure and the one
+    # inside `get_builder` are different
+    assert args['space']['structure'].uuid != in_struct.uuid
+
+    in_struct = generate_structure(symbols=('Si',)).store()
+    buil, args = class_ins.get_builder(space={'structure': in_struct})  # pylint: disable=unused-variable
+    # in_struct is now stored, therefore it is maintained
+    # within `get_builder`
+    assert args['space']['structure'].uuid == in_struct.uuid
```
Don't need to explicitly create the `space` namespace if you define a port in it, it will be created automatically on the way.
```suggestion
    cls = generate_input_generator_cls(inputs_dict={'space.structure': orm.StructureData})
    generator = cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))

    structure = generate_structure(symbols=('Si',))
    kwargs = {'space': {'structure': structure}}
    generator.get_builder(**kwargs)
    # structure is an unstored node, therefore it is deepcopied and so the UUID should change.
    assert kwargs['space']['structure'].uuid != structure.uuid

    structure = generate_structure(symbols=('Si',)).store()
    kwargs = {'space': {'structure': structure}}
    generator.get_builder(**kwargs)
    # The structure is now stored and so should be the same.
    assert kwargs['space']['structure'].uuid == structure.uuid
```

sphuber reviewed (changes requested):
Thanks @bosonie . Implementation looks good. Not sure if we actually need to distinguish between stored and unstored nodes, but can't hurt for now. Anyway, they would be stored before the `get_builder` is called when going through the compound workflows. Just have some changes in naming and general cleaning up, but then should be good to go.

sphuber commented on aiida_common_workflows/generators/generator.py:
```diff
@@ -3,13 +3,31 @@
+    """
+    if isinstance(obj, dict):
+        return {k: recursively_check_stored_nodes(v) for k, v in obj.items()}
+    if isinstance(obj, orm.Data) and obj.is_stored:
```
Actually, I think we should not limit this to just `Data` nodes. It is possible that `ProcessNodes` are passed as well, for example the `reference_workchain`, and there `deepcopy` will even raise. So we should also just return it.

bosonie commented:
> Not sure if we actually need to distinguish between stored and unstored nodes, but can't hurt for now. Anyway, they would be stored before the `get_builder` is called when going through the compound workflows.

This is a valid point. In other words, it does not make sense to distinguish stored and unstored nodes since it can always be that the node is stored before going through the `get_builder` (for instance if the `get_builder` call is inside a higher level workchain exposing the node as input). But this also means that we should remember all the developers that they should not modify ever a node inside the `get_builder`. It should be already clear but better remember that. However cloning or deepcopying internally is allowed even though the connection among the original node and the cloned one is lost since it is not in a `calcfunction`.

bosonie commented:
@sphuber I finished to implement your modifications. Please note that:
1) I extended the imposition to NOT deepcopy to all `Node`s, no metter they are stored or not.
2) I modified the tests to test that a `Node` is not deepcopied and a `dict` is deepcopied. This is done adding that the `dict` is modified inside `get_builder`, but the modification has no effect in the passed kwarg. For the  `Node` only the uuid is checked. Is this enough?

sphuber commented on tests/generators/test_generator.py:
```diff
@@ -0,0 +1,63 @@
+
+    cls = generate_input_generator_cls(inputs_dict={'mutable': dict})
+    generator = cls(process_class=WorkflowFactory('common_workflows.relax.siesta'))
+    kwargs = {'mutable': {'test': 333}}
+    generator.get_builder(**kwargs)
+    assert kwargs['mutable'] == {'test': 333}
```
It took me a while to understand why this test was actually testing something. It relies on the `_construct_builder` implementation of the class that is returned by the `generate_input_generator_cls` factory. This is not at all clear and the docstring also doesn't say this. For this purpose, I would really advice to have the implementation of `_construct_builder` locally (either in this test file or better even in the test itself). You don't have to copy the entire class and implement it again, you can just monkeypatch the relevant method. I would suggest doing the following:
```python
def test_get_builder_mutable_kwargs(generate_input_generator_cls, monkeypatch):
    
    def construct_builder(self, **kwargs):
        kwargs['mutable']['test'] = 'whatever'
        return self.process_class.get_builder()

    cls = generate_input_generator_cls(inputs_dict={'mutable': dict})
    monkeypatch.setattr(cls, '_construct_builder, construct_builder)
    # and then do the rest of the test.
```

This is a lot clearer and also less susceptible to accidentally breaking the test if the fixture `generate_input_generator_cls` removes the logic that changes the arbitrary keys in the `kwargs` that is not clear at all why it is there.

sphuber commented on tests/conftest.py:
```diff
@@ -63,6 +64,43 @@ def _run_cli_command(command: click.Command, options: list = None, raises: bool
+                        spec.input(k, valid_type=val)
+
+            def _construct_builder(self, **kwargs) -> engine.ProcessBuilder:
+                if 'mutable' in kwargs:
+                    kwargs['mutable']['test'] = 'whatever'
```
would remove this. It is specific to one test but really difficult to figure out which one. Should be locally implemented. See other comment

sphuber commented on aiida_common_workflows/generators/generator.py:
```diff
@@ -58,7 +72,15 @@ def get_builder(self, **kwargs) -> engine.ProcessBuilder:
         the ``parsed_kwargs`` attribute.
         """
-        processed_kwargs = self.spec().inputs.pre_process(copy.deepcopy(kwargs))
+        #Create a deep copy of the input arguments because the ``pre_process`` step may alter them and
+        #the originals need to be preserved in case they are passed to the ``get_builder`` method again, as
+        #for example within a loop of a code-agnostic wrapping workchain like the ``EquationOfStateWorkChain``.
+        #We cannot use ``copy.deepcopy`` directly, however, since it will create clones of any stored nodes that
+        #are in the inputs. That's why we call `recursively_check_stored_nodes` which will recursively create a
+        #deep copy of all inputs except for the stored nodes.
```
Don't understand why the linter didn't pick this up...
```suggestion
        # Create a deep copy of the input arguments because the ``pre_process`` step may alter them and
        # the originals need to be preserved in case they are passed to the ``get_builder`` method again, as
        # for example within a loop of a code-agnostic wrapping workchain like the ``EquationOfStateWorkChain``.
        # We cannot use ``copy.deepcopy`` directly, however, since it will create clones of any stored nodes that
        # are in the inputs. That's why we call `recursively_check_stored_nodes` which will recursively create a
        # deep copy of all inputs except for the stored nodes.
```

sphuber reviewed (changes requested):
Thanks @bosonie , just few remaining things

bosonie commented:
@sphuber Thanks for the comments. I hope I clarified all the points.

sphuber commented:
Thanks @bosonie
